Template - CSV file to JSON file

jannejjj opened the pull request:
I previously thought that there was a bug in the SFTP.WriteFile task similar to the one in the UploadFiles, but it turns out I was just using an incorrect file path. Everything works as it should. 😄

closes #8

Checklist for review:
- [x] Process name is fine, uses normal casing and descriptive
- [x] Process description is fine, uses normal casing, dots and is descriptive
  - [x] Element names use normal casing, no unneeded capital letters.
- [x] Correct casing used for formats and terms, e.g. CSV, FTP, JSON, Avro
  - [x] Element names do not contain dots
  - [x] Comments / note element text are in good language and end with dots.
- [x] Template parameters checked
  - [x] Parameter descriptions are good and end with dots.
  - [x] Sensitive parameters are marked as secrets
- [x] "Turn off logging" switch is on for any task that has sensitive data as output
- [x] End "balls" like error throws and return statements have titles (double-click) and reasonable information.
  - [x] Results have good summaries, preferably with statistics of some sort like "processed X files"
  - [x] Errors have some error info, usually from task (e.g. some ErrorMessage property contains additional error info)
- [x] Process is organised with groups (dashed boxes)
- [x] Straight lines are preferred where possible
- [x] Subprocesses are forbidden
- [x] Text mode is preferred over Expressions in expressions, e.g. instead of *$"This is {#var.error}"* use *This is {{#var.error}}*
- [x] Process metadata file is present and content is fine (version number and name)
- [x] Process export file is present (process.json)
- [x] Process description file is present (all sections filled in)
- [x] Process description has a SVG screenshot of the process
- [x] Tempalte files are going to the correct folder


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a new template that downloads a CSV file from an SFTP server, converts it to JSON format, and uploads the result back to the SFTP server.
  - Added metadata for the "CSV file to JSON file" transformation, including name, version, tags, and additional systems used.
  - Detailed process descriptions, variable definitions, and logic flow in BPMN format for the CSV to JSON file conversion process.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

FrendsSarlinS reviewed (commented):
Checklist for review:
- [ ] Process name is fine, uses normal casing and descriptive
- [ ] Process description is fine, uses normal casing, dots and is descriptive
  - [ ] Element names use normal casing, no unneeded capital letters.
- [ ] Correct casing used for formats and terms, e.g. CSV, FTP, JSON, Avro
  - [ ] Element names do not contain dots
  - [ ] Comments / note element text are in good language and end with dots.
- [ ] Template parameters checked
  - [ ] Parameter descriptions are good and end with dots.
  - [ ] Sensitive parameters are marked as secrets
- [ ] "Turn off logging" switch is on for any task that has sensitive data as output
- [ ] End "balls" like error throws and return statements have titles (double-click) and reasonable information.
  - [ ] Results have good summaries, preferably with statistics of some sort like "processed X files"
  - [ ] Errors have some error info, usually from task (e.g. some ErrorMessage property contains additional error info)
- [ ] Process is organized with groups (dashed boxes)
- [ ] Straight lines are preferred where possible
- [ ] Subprocesses are forbidden
- [ ] Text mode is preferred over Expressions in expressions, e.g. instead of *$"This is {#var.error}"* use *This is {{#var.error}}*
- [ ] Process metadata file is present and content is fine (version number and name)
- [ ] Process export file is present (process.json)
- [ ] Process description file is present (all sections filled in)
- [ ] Process description has a SVG screenshot of the process
- [ ] Template files are going to the correct folder

FrendsSarlinS commented:
1) The process and the template name should be "CSV file to JSON file" (uppercase JSON) to be in accord with other processes and templates that have to do with JSON. Right now it's "Json" for both process and template.

2) However, in metadata.json, the title is "CSV file to JSON file" as it should.

3) Description for the process has the same issue with JSON casing. Also, the grammar is not correct.

4) Process variable "TargetFileName" is missing an ending dot, or period, in the description.

5) Is it possible to get the task error message as an output as well for both Throws? Right now, the error messages seem lacking.

jannejjj commented:
> 1. The process and the template name should be "CSV file to JSON file" (uppercase JSON) to be in accord with other processes and templates that have to do with JSON. Right now it's "Json" for both process and template.

Changed casing.

> 3. Description for the process has the same issue with JSON casing. Also, the grammar is not correct.

I don't think the _process_ description matters, since that doesn't actually go into the template when you make a template out of the process. Changed it anyways. :D

> 4. Process variable "TargetFileName" is missing an ending dot, or period, in the description.

Added the missing period.

> 5. Is it possible to get the task error message as an output as well for both Throws? Right now, the error messages seem lacking.

Sadly, the SFTP read and upload tasks throw exceptions instead of returning error messages, meaning that actually, if the task runs into an error, the process simply stops there:

![image](https://github.com/FrendsPlatform/FrendsTemplates/assets/61980297/013f23a8-7ce6-4ded-b24e-d095e1da665f)

jannejjj commented on Templates/Files/SFTP/CSV file to JSON file/process.json:
```diff
@@ -0,0 +1,79 @@
+			"Modified": "2024-06-27T07:38:29.6144707",
+			"Modifier": "janne.pennanen@frends.com",
+			"Tags": [
+				"Json",
```
Fixed!

FrendsSarlinS reviewed (approved):
After the changes, the process looks good to me! Ticked all the checkboxes.

jefim commented on Templates/Files/SFTP/CSV file to JSON file/assets/CSV_file_to_JSON_file.svg:
1. Throw error if file read was not successful - add reason why read was not successful.
2. Write JSON file to SFTP server - use `{{#result.Json}}` in text mode instead of ToString.
3. Was the write successful? - if this goes to this check already, then results are never null anyway, so the check for null is redundant.
4. Throw error if write was not successful - specify why we think that write was not successful

jannejjj replied:
Thanks for the comments! Added mentions of empty files to the throws, changed from ToString to handlebars and text mode and removed the null check.

FrendsSarlinS reviewed (approved):
Re-checked this template, everything works fine and checks all boxes. The issues posted by Jefim were also addressed.
